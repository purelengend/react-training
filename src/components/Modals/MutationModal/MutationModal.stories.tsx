import type { Meta, StoryObj } from '@storybook/react';
import MutationModal from './index';
import './mutation-modal.module.css';
const meta: Meta<typeof MutationModal> = {
  title: 'Example/MutationModal',
  component: MutationModal,
  argTypes: {
    isVisible: {
      description: 'Indicates whether the modal is hidden or visible.'
    },
    title: {
      description: 'The title of the modal'
    },
    prodData: {
      description: 'The default data for the modal'
    }
  }
};

type Story = StoryObj<typeof MutationModal>;

export const MutationModalExample: Story = {
  args: {
    isVisible: true,
    title: 'Sample Title',
    prodData: {
      id: '0',
      name: 'Sample name',
      price: 213,
      imageUrl:
        'https://images.unsplash.com/photo-1614777986387-015c2a89b696?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=3774&q=80',
      quantity: 23,
      createdAt: new Date().toDateString()
    }
  }
};
export default meta;
