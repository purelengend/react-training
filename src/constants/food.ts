import { Food } from '@components/common/Cards/ProductCard';

export const FOOD_NAME_WARNING_MSG =
  'The food name cannot be empty or have two trailing spaces consecutive.';
export const FOOD_PRICE_WARNING_MSG =
  'The food price cannot be empty and greater than 0';
export const FOOD_IMG_WARNING_MSG =
  'The food image URL cannot be empty and must be in the right format';
export const FOOD_QUANTITY_WARNING_MSG =
  'The food quantity cannot be empty and must be an integer that greater than 0';
export const EMPTY_MSG =
  'Oops! It looks like there are no food items available at the moment.';
export const DEFAULT_FOOD_ID_VALUE = '';

export const defaultData: Food = {
  id: '',
  name: '',
  price: 0,
  imageUrl: '',
  quantity: 0,
  createdAt: new Date()
};

export interface FoodErrorMessage {
  name: string;
  price: string;
  imageUrl: string;
  quantity: string;
}

export const defaultFoodErrorMessage: FoodErrorMessage = {
  name: '',
  price: '',
  imageUrl: '',
  quantity: ''
};
