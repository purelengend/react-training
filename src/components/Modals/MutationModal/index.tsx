import mutationModalStyles from '@components/Modals/MutationModal/mutation-modal.module.css';
import { Button } from '@components/common/Button';
import { Food } from '@components/common/Cards/ProductCard';
import { InputField } from '@components/common/InputField';
import {
  FormEvent,
  memo,
  useCallback,
  useContext,
  useEffect,
  useState
} from 'react';
import { ModalContext } from '@context/modal';
import { validateForm } from '@helpers/form-validation';
import { defaultData } from '@constants/food';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { mutationFood } from '@services/food.service';
import {
  TOAST_ADD_MSG,
  TOAST_EDIT_MSG,
  TOAST_ERROR_MSG,
  TOAST_TIME
} from '@constants/toast';
import { InfiniteQueryProps } from '@hooks/useFood';
import { deepClone } from '@helpers/deep-clone';

interface MutationModalProps {
  isVisible: boolean;
  title: string;
  prodData?: Food;
}

export interface FoodErrorMessage {
  name: string;
  price: string;
  imageUrl: string;
  quantity: string;
}

const defaultErrorMessage: FoodErrorMessage = {
  name: '',
  price: '',
  imageUrl: '',
  quantity: ''
};
const MutationModal = memo(
  ({ isVisible, title, prodData = defaultData }: MutationModalProps) => {
    const [mutationData, setMutationData] = useState(prodData);
    const [errorMessage, setErrorMessage] = useState(defaultErrorMessage);
    const { setMutationShowUp, setLoadingShowUp, showToast, hideToast } =
      useContext(ModalContext);
    const queryClient = useQueryClient();

    useEffect(() => {
      setMutationData(prodData);
    }, [prodData]);

    const mutation = useMutation({
      mutationFn: (input: Food) => {
        return mutationFood(input);
      },
      onMutate: () => {
        setLoadingShowUp(true);
      },
      onSuccess: data => {
        const currentFoodData = queryClient.getQueryData<
          InfiniteQueryProps<Food>
        >(['foods']);
        let toastMessage = '';
        if (currentFoodData) {
          let existedFoodIndex = -1;

          // Loop all food pages, check the data prop and loop over all food items in data to find the existed food
          for (const [index, foodPage] of currentFoodData.pages.entries()) {
            const foundedFoodIndex = foodPage.data.findIndex(food => {
              return food.id === data.id;
            });

            // If the food is exist, update the local client food data
            if (foundedFoodIndex > -1) {
              existedFoodIndex = foundedFoodIndex;

              const updatedFoodData =
                deepClone<InfiniteQueryProps<Food>>(currentFoodData);

              updatedFoodData.pages[index].data[foundedFoodIndex] = data;
              queryClient.setQueryData<InfiniteQueryProps<Food>>(
                ['foods'],
                updatedFoodData
              );
            }
          }

          // Otherwise, re-call the get food API
          if (existedFoodIndex < 0) {
            toastMessage = TOAST_ADD_MSG;
            queryClient.resetQueries({ queryKey: ['foods'] });
          } else {
            toastMessage = TOAST_EDIT_MSG;
          }
        }
        onCancelClick();
        setLoadingShowUp(false);
        showToast(toastMessage, true);
        setTimeout(() => {
          hideToast();
        }, TOAST_TIME);
      },
      onError: () => {
        onCancelClick();
        setLoadingShowUp(false);
        showToast(TOAST_ERROR_MSG, false);
        setTimeout(() => {
          hideToast();
        }, TOAST_TIME);
      }
    });

    const onChangeMutation = (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      setMutationData({
        ...mutationData,
        [e.target.name]: value
      });
    };

    const onCancelClick = useCallback(() => {
      if (mutationData.id === defaultData.id) setMutationData(defaultData);
      else setMutationData(prodData);
      setErrorMessage(defaultErrorMessage);
      setMutationShowUp(false);
    }, [mutationData.id, prodData, setMutationShowUp]);

    const onSubmit = (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      const validateMessage = validateForm(mutationData);
      if (Object.values(validateMessage).join('')) {
        setErrorMessage(validateMessage);
      } else {
        mutation.mutate(mutationData);
      }
    };
    return (
      isVisible && (
        <div id="mutation-modal" className={`d-flex-center modal-overlay`}>
          <div
            className={`d-flex-col ${mutationModalStyles['mutation-modal-body']}`}
          >
            <h2
              id="mutation-title"
              className={mutationModalStyles['mutation-modal-content']}
            >
              {title}
            </h2>
            <form
              id="mutation-form"
              className={`d-flex-col ${mutationModalStyles['mutation-form']}`}
              onSubmit={onSubmit}
            >
              <input
                type="hidden"
                id="food-id"
                name="food-id"
                value={mutationData.id}
              />
              <input
                type="hidden"
                id="created-at"
                name="created-at"
                value={
                  typeof mutationData.createdAt === 'string'
                    ? mutationData.createdAt
                    : mutationData.createdAt.toDateString()
                }
              />
              <div
                className={`d-flex-col ${mutationModalStyles['mutation-form-field']}`}
              >
                <InputField
                  label="Name"
                  htmlFor="name"
                  labelClass={mutationModalStyles['mutation-label']}
                  type="text"
                  inputClass={mutationModalStyles['mutation-input']}
                  name="name"
                  value={mutationData.name}
                  onChange={onChangeMutation}
                />
                {errorMessage.name && (
                  <p className={mutationModalStyles['mutation-warning']}>
                    {errorMessage.name}
                  </p>
                )}
              </div>
              <div
                className={`d-flex-col ${mutationModalStyles['mutation-form-field']}`}
              >
                <InputField
                  label="Price"
                  htmlFor="price"
                  labelClass={mutationModalStyles['mutation-label']}
                  type="number"
                  inputClass={mutationModalStyles['mutation-input']}
                  name="price"
                  value={`${mutationData.price}`}
                  onChange={onChangeMutation}
                />
                {errorMessage.price && (
                  <p className={mutationModalStyles['mutation-warning']}>
                    {errorMessage.price}
                  </p>
                )}
              </div>
              <div
                className={`d-flex-col ${mutationModalStyles['mutation-form-field']}`}
              >
                <InputField
                  label="Image URL"
                  htmlFor="image"
                  labelClass={mutationModalStyles['mutation-label']}
                  type="text"
                  inputClass={mutationModalStyles['mutation-input']}
                  name="imageUrl"
                  value={mutationData.imageUrl}
                  onChange={onChangeMutation}
                />
                {errorMessage.imageUrl && (
                  <p className={mutationModalStyles['mutation-warning']}>
                    {errorMessage.imageUrl}
                  </p>
                )}
              </div>
              <div
                className={`d-flex-col ${mutationModalStyles['mutation-form-field']}`}
              >
                <InputField
                  label="Quantity"
                  htmlFor="quantity"
                  labelClass={mutationModalStyles['mutation-label']}
                  type="number"
                  inputClass={`${mutationModalStyles['mutation-input']} ${mutationModalStyles.half}`}
                  name="quantity"
                  value={`${mutationData.quantity}`}
                  onChange={onChangeMutation}
                />
                {errorMessage.quantity && (
                  <p className={mutationModalStyles['mutation-warning']}>
                    {errorMessage.quantity}
                  </p>
                )}
              </div>
              <div
                className={`d-flex ${mutationModalStyles['mutation-modal-btn-wrapper']}`}
              >
                <Button onClick={onCancelClick} className="modal-btn cancel">
                  Cancel
                </Button>
                <Button type="submit" className="modal-btn confirm">
                  Save
                </Button>
              </div>
            </form>
          </div>
        </div>
      )
    );
  }
);

MutationModal.whyDidYouRender = true;
export default MutationModal;
