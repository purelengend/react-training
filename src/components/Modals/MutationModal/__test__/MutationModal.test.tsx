import { defaultData, defaultFoodErrorMessage } from '@constants/food';
import { customRenderer, render, screen } from '@src/test/test-utils';
import userEvent from '@testing-library/user-event';

import MutationModal, { MutationModalProps } from '..';

describe('MutationModal test case', () => {
  const mockMutationModalProps: MutationModalProps = {
    title: 'mock',
    productData: defaultData,
    errorProductMessage: defaultFoodErrorMessage,
    onSubmit: vi.fn(),
    onCancelClick: vi.fn(),
    setProductData: vi.fn()
  };

  it('should render correctly', () => {
    const mockMutationModal = customRenderer(
      <MutationModal {...mockMutationModalProps} />
    );

    expect(mockMutationModal.toJSON()).toMatchSnapshot();
  });

  it('should invoke onSubmit function when submitting', async () => {
    render(<MutationModal {...mockMutationModalProps} />);

    const mockButtonList = screen.getAllByRole('button');

    await userEvent.click(mockButtonList[1]);

    expect(mockMutationModalProps.onSubmit).toBeCalled();
  });

  it('should invoke onCancelClick function when closing the modal', async () => {
    render(<MutationModal {...mockMutationModalProps} />);

    const mockButtonList = screen.getAllByRole('button');

    await userEvent.click(mockButtonList[0]);

    expect(mockMutationModalProps.onCancelClick).toBeCalled();
  });

  it('should invoke setProductData function when filling the modal', async () => {
    render(<MutationModal {...mockMutationModalProps} />);

    const mockInputList = screen.getAllByRole('textbox');

    await userEvent.type(mockInputList[0], 'mock');

    expect(mockMutationModalProps.setProductData).toBeCalled();
  });
});
