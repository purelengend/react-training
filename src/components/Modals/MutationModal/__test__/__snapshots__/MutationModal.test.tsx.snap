// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MutationModal test case > should render correctly 1`] = `
<div
  className="d-flex-center modal-overlay"
  id="mutation-modal"
>
  <div
    className="d-flex-col _mutation-modal-body_9aa3b9"
  >
    <h2
      className="_mutation-modal-content_9aa3b9"
      id="mutation-title"
    >
      mock
    </h2>
    <form
      className="d-flex-col _mutation-form_9aa3b9"
      id="mutation-form"
      onSubmit={[MockFunction spy]}
    >
      <div
        className="d-flex-col _mutation-form-field_9aa3b9"
      >
        <label
          className="_mutation-label_9aa3b9"
          htmlFor="name"
        >
          Name
        </label>
        <input
          className="_mutation-input_9aa3b9"
          id="name"
          name="name"
          onChange={[Function]}
          step="any"
          type="text"
          value=""
        />
      </div>
      <div
        className="d-flex-col _mutation-form-field_9aa3b9"
      >
        <label
          className="_mutation-label_9aa3b9"
          htmlFor="price"
        >
          Price
        </label>
        <input
          className="_mutation-input_9aa3b9"
          id="price"
          name="price"
          onChange={[Function]}
          step="any"
          type="number"
          value="0"
        />
      </div>
      <div
        className="d-flex-col _mutation-form-field_9aa3b9"
      >
        <label
          className="_mutation-label_9aa3b9"
          htmlFor="image"
        >
          Image URL
        </label>
        <input
          className="_mutation-input_9aa3b9"
          id="image"
          name="imageUrl"
          onChange={[Function]}
          step="any"
          type="text"
          value=""
        />
      </div>
      <div
        className="d-flex-col _mutation-form-field_9aa3b9"
      >
        <label
          className="_mutation-label_9aa3b9"
          htmlFor="quantity"
        >
          Quantity
        </label>
        <input
          className="_mutation-input_9aa3b9 _half_9aa3b9"
          id="quantity"
          name="quantity"
          onChange={[Function]}
          step="any"
          type="number"
          value="0"
        />
      </div>
      <div
        className="d-flex _mutation-modal-btn-wrapper_9aa3b9"
      >
        <button
          className="modal-btn cancel"
          data-id="0"
          disabled={false}
          onClick={[MockFunction spy]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="modal-btn confirm"
          data-id="0"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
